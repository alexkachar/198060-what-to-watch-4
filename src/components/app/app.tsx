import * as React from 'react';
import {BrowserRouter, Route, Switch} from 'react-router-dom';
import {connect} from 'react-redux';

import {AppRoutes} from '../../constants';
import {filterMovies, getMoviesLimit} from '../../store/reducers/ui/selectors';
import {getPromoMovie, getGenres, getLoadingFlag} from '../../store/reducers/data/selectors';
import {getSelectedGenre} from '../../store/reducers/ui/selectors';
import {getAuthFlag} from '../../store/reducers/user/selectors';
import {MOVIES_LIMIT_ADD_STEP} from '../../constants';
import UiActionCreator from '../../store/actions/ui/ui';
import UserOperation from '../../store/operations/user/user';
import DataOperation from '../../store/operations/data/data';

import Movie from '../../interfaces/movie';
import Main from '../main/main';
import MoviePage from '../movie-page/movie-page';
import Login from '../login/login';
import Favorites from '../favorites/favorites';
import ReviewForm from '../review-form/review-form';

interface Props {
  movies: Movie[];
  promoMovie: Movie;
  genres: string[];
  selectedGenre: string;
  moviesLimit: number;
  loading: boolean;
  isAuth: boolean;
  onGenreSelect: (genre: string) => void;
  onShowMoreClick: () => void;
  onLogin: (authData: {}) => void;
  onSetFavoriteStatus: (movieId: number | string, isFavorite: boolean) => void;
}

const App = (props: Props) => {
  const {
    isAuth,
    movies,
    promoMovie,
    genres,
    selectedGenre,
    moviesLimit,
    loading,
    onGenreSelect,
    onShowMoreClick,
    onLogin,
    onSetFavoriteStatus,
  } = props;

  const showMoreAccess = movies.length > moviesLimit;
  const moviesToList = movies.slice(0, moviesLimit);
  return (
    <BrowserRouter>
      <Switch>
        <Route exact path={AppRoutes.MAIN}
          render={() => {
            return (
              <Main
                movies={moviesToList}
                promoMovie={promoMovie}
                genres={genres}
                selectedGenre={selectedGenre}
                showMoreAccess={showMoreAccess}
                onGenreSelect={onGenreSelect}
                onShowMoreClick={onShowMoreClick}
                onSetFavoriteStatus={onSetFavoriteStatus}
                loading={loading}
                isAuth={isAuth}
              />
            );
          }}
        />

        <Route exact path={AppRoutes.MOVIE}
          render={({match}) => {
            const {id} = match.params;
            return (
              <MoviePage
                movieId={id}
                onSetFavoriteStatus={onSetFavoriteStatus}
              />
            );
          }}
        />

        <Route exact path={AppRoutes.ADD_REVIEW}
          render={({match}) => {
            const {id} = match.params;
            return (
              <ReviewForm
                movieId={id}
                isAuth={isAuth}
              />
            );
          }}
        />

        <Route exact path={AppRoutes.LOGIN}
          render={() => {
            return (
              <Login
                onLogin={onLogin}
                loading={loading}
                isAuth={isAuth}
              />
            );
          }}
        />

        <Route exact path={AppRoutes.FAVORITES}
          render={() => {
            return (
              <Favorites />
            );
          }}
        />
      </Switch>
    </BrowserRouter>
  );
};

const mapStateToProps = (state) => ({
  movies: filterMovies(state),
  promoMovie: getPromoMovie(state),
  genres: getGenres(state),
  selectedGenre: getSelectedGenre(state),
  moviesLimit: getMoviesLimit(state),
  loading: getLoadingFlag(state),
  isAuth: getAuthFlag(state)
});

const mapDispatchToProps = (dispatch) => ({

  onGenreSelect(genre: string) {
    dispatch(UiActionCreator.selectGenre(genre));
    dispatch(UiActionCreator.resetMoviesLimit());
  },

  onShowMoreClick: () => {
    dispatch(UiActionCreator.setMoviesLimit(MOVIES_LIMIT_ADD_STEP));
  },

  onLogin(authData: {}) {
    dispatch(UserOperation.login(authData));
  },

  onSetFavoriteStatus(movieId, isFavorite: boolean) {
    dispatch(DataOperation.setFavoriteStatus(movieId, isFavorite));
  },

});

export {App};
export default connect(mapStateToProps, mapDispatchToProps)(App);
