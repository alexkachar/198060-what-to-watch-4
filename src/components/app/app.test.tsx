
import * as React from 'react';
import * as renderer from 'react-test-renderer';
import {App} from './app';
import MOCK_MOVIES from '../../test-data/mock-movies';
import MOCK_GENRES from '../../test-data/mock-genres';

it(`App renders correctly`, () => {
  const tree = renderer
    .create(
        <App
          movies={MOCK_MOVIES}
          genres={MOCK_GENRES}
          onGenreSelect={jest.fn()}
        />, {
          createNodeMock: () => document.createElement(`div`)
        })
    .toJSON();
  expect(tree).toMatchSnapshot();
});
