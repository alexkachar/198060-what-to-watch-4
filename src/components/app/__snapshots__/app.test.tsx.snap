// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders correctly 1`] = `
<div>
  <section
    className="movie-card"
  >
    <div
      className="movie-card__bg"
    >
      <img
        alt="The Grand Budapest Hotel"
        src="img/the-grand-budapest-hotel-bg.jpg"
      />
    </div>
    <h1
      className="visually-hidden"
    >
      WTW
    </h1>
    <header
      className="page-header movie-card__head"
    >
      <div
        className="logo"
      >
        <a
          className="logo__link"
          href="/"
          onClick={[Function]}
        >
          <span
            className="logo__letter logo__letter--1"
          >
            W
          </span>
          <span
            className="logo__letter logo__letter--2"
          >
            T
          </span>
          <span
            className="logo__letter logo__letter--3"
          >
            W
          </span>
        </a>
      </div>
      <div
        className="user-block"
      >
        <a
          className="user-block__avatar"
          href="/favorites"
          onClick={[Function]}
          style={
            Object {
              "textDecoration": "none",
            }
          }
        >
          <img
            src="https://htmlacademy-react-3.appspot.com/img/john_doe.jpg"
            style={
              Object {
                "borderRadius": "50%",
                "height": "63px",
                "width": "63px",
              }
            }
          />
        </a>
      </div>
    </header>
    <div
      className="movie-card__wrap"
    >
      <div
        className="movie-card__info"
      >
        <div
          className="movie-card__poster"
        >
          <img
            alt="The Grand Budapest Hotel poster"
            height={327}
            src="img/the-grand-budapest-hotel-poster.jpg"
            width={218}
          />
        </div>
        <div
          className="movie-card__desc"
        >
          <h2
            className="movie-card__title"
          >
            The Grand Budapest Hotel
          </h2>
          <p
            className="movie-card__meta"
          >
            <span
              className="movie-card__genre"
            >
              Comedy
            </span>
            <span
              className="movie-card__year"
            >
              2014
            </span>
          </p>
          <div
            className="movie-card__buttons"
          >
            <button
              className="btn btn--play movie-card__button"
              type="button"
            >
              <svg
                height={19}
                viewBox="0 0 21 19"
                width={21}
              >
                <use
                  xlinkHref="#play-s"
                />
              </svg>
              <span>
                Play
              </span>
            </button>
            <button
              className="btn btn--list movie-card__button"
              type="button"
            >
              <svg
                height={20}
                viewBox="0 0 19 20"
                width={19}
              >
                <use
                  xlinkHref="#add"
                />
              </svg>
              <span>
                My list
              </span>
            </button>
          </div>
        </div>
      </div>
    </div>
  </section>
  <div
    className="page-content"
  >
    <section
      className="catalog"
    >
      <h2
        className="catalog__title visually-hidden"
      >
        Catalog
      </h2>
      <ul
        className="catalog__genres-list"
      >
        <li
          className="catalog__genres-item catalog__genres-item--active"
        >
          <a
            className="catalog__genres-link"
            onClick={[Function]}
            style={
              Object {
                "cursor": "pointer",
              }
            }
          >
            All genres
          </a>
        </li>
        <li
          className="catalog__genres-item null"
        >
          <a
            className="catalog__genres-link"
            onClick={[Function]}
            style={
              Object {
                "cursor": "pointer",
              }
            }
          >
            Adventure
          </a>
        </li>
        <li
          className="catalog__genres-item null"
        >
          <a
            className="catalog__genres-link"
            onClick={[Function]}
            style={
              Object {
                "cursor": "pointer",
              }
            }
          >
            Comedy
          </a>
        </li>
        <li
          className="catalog__genres-item null"
        >
          <a
            className="catalog__genres-link"
            onClick={[Function]}
            style={
              Object {
                "cursor": "pointer",
              }
            }
          >
            Fantasy
          </a>
        </li>
      </ul>
      <div
        className="catalog__movies-list"
      >
        <article
          className="small-movie-card catalog__movies-card"
        >
          <a
            href="/film/1"
            onClick={[Function]}
          >
            <div
              className="small-movie-card__image"
            >
              <img
                alt="The Grand Budapest Hotel"
                height={175}
                src="img/the-grand-budapest-hotel.jpg"
                width={280}
              />
            </div>
          </a>
          <h3
            className="small-movie-card__title"
          >
            <a
              className="small-movie-card__link"
              href="/film/1"
              onClick={[Function]}
              style={
                Object {
                  "color": "#c9b37e",
                }
              }
            >
              The Grand Budapest Hotel
            </a>
          </h3>
        </article>
        <article
          className="small-movie-card catalog__movies-card"
        >
          <a
            href="/film/2"
            onClick={[Function]}
          >
            <div
              className="small-movie-card__image"
            >
              <img
                alt="Moonrise Kingdom"
                height={175}
                src="https://htmlacademy-react-3.appspot.com/wtw/static/film/preview/moonrise-kingdom.jpg"
                width={280}
              />
            </div>
          </a>
          <h3
            className="small-movie-card__title"
          >
            <a
              className="small-movie-card__link"
              href="/film/2"
              onClick={[Function]}
              style={
                Object {
                  "color": "#c9b37e",
                }
              }
            >
              Moonrise Kingdom
            </a>
          </h3>
        </article>
        <article
          className="small-movie-card catalog__movies-card"
        >
          <a
            href="/film/15"
            onClick={[Function]}
          >
            <div
              className="small-movie-card__image"
            >
              <img
                alt="Fantastic Beasts: The Crimes of Grindelwald"
                height={175}
                src="https://htmlacademy-react-3.appspot.com/wtw/static/film/preview/fantastic-beasts-the-crimes-of-grindelwald.jpg"
                width={280}
              />
            </div>
          </a>
          <h3
            className="small-movie-card__title"
          >
            <a
              className="small-movie-card__link"
              href="/film/15"
              onClick={[Function]}
              style={
                Object {
                  "color": "#c9b37e",
                }
              }
            >
              Fantastic Beasts: The Crimes of Grindelwald
            </a>
          </h3>
        </article>
      </div>
    </section>
    <footer
      className="page-footer"
    >
      <div
        className="logo"
      >
        <a
          className="logo__link logo__link--light"
          href="/"
          onClick={[Function]}
        >
          <span
            className="logo__letter logo__letter--1"
          >
            W
          </span>
          <span
            className="logo__letter logo__letter--2"
          >
            T
          </span>
          <span
            className="logo__letter logo__letter--3"
          >
            W
          </span>
        </a>
      </div>
      <div
        className="copyright"
      >
        <p>
          Â© 2019 What to watch Ltd.
        </p>
      </div>
    </footer>
  </div>
</div>
`;
